README.md
# oocyte_maturity_classifier

CHANGELOG.md
## 0.0.0 (September 28, 2023)
  - bump-version.sh added
  - Merge pull request #1 from jesuslife/feature/setup_repo
  - Data and Weights folders pushed to bucket
  - data and weights folders added to DVC
  - Initial commit



Code/hubconf.py
# YOLOv5 🚀 by Ultralytics, GPL-3.0 license
"""
PyTorch Hub models https://pytorch.org/hub/ultralytics_yolov5

Usage:
    import torch
    model = torch.hub.load('ultralytics/yolov5', 'yolov5s')  # official model
    model = torch.hub.load('ultralytics/yolov5:master', 'yolov5s')  # from branch
    model = torch.hub.load('ultralytics/yolov5', 'custom', 'yolov5s.pt')  # custom/local model
    model = torch.hub.load('.', 'custom', 'yolov5s.pt', source='local')  # local repo
"""

import torch


def _create(name, pretrained=True, channels=3, classes=80, autoshape=True, verbose=True, device=None):
    """Creates or loads a YOLOv5 model

    Arguments:
        name (str): model name 'yolov5s' or path 'path/to/best.pt'
        pretrained (bool): load pretrained weights into the model
        channels (int): number of input channels
        classes (int): number of model n_classes
        autoshape (bool): apply YOLOv5 .autoshape() wrapper to model
        verbose (bool): print all information to screen
        device (str, torch.device, None): device to use for model parameters

    Returns:
        YOLOv5 model
    """
    from pathlib import Path

    from yolov5.models.common import AutoShape, DetectMultiBackend
    from yolov5.models.experimental import attempt_load
    from yolov5.models.yolo import ClassificationModel, DetectionModel, SegmentationModel
    from yolov5.utils.downloads import attempt_download
    from yolov5.utils.general import LOGGER, check_requirements, intersect_dicts, logging
    from yolov5.utils.torch_utils import select_device

    if not verbose:
        LOGGER.setLevel(logging.WARNING)
    check_requirements(exclude=('opencv-python', 'tensorboard', 'thop'))
    name = Path(name)
    path = name.with_suffix('.pt') if name.suffix == '' and not name.is_dir() else name  # checkpoint path
    try:
        device = select_device(device)
        if pretrained and channels == 3 and classes == 80:
            try:
                model = DetectMultiBackend(path, device=device, fuse=autoshape)  # detection model
                if autoshape:
                    if model.pt and isinstance(model.model, ClassificationModel):
                        LOGGER.warning('WARNING ⚠️ YOLOv5 ClassificationModel is not yet AutoShape compatible. '
                                       'You must pass torch tensors in BCHW to this model, i.e. shape(1,3,224,224).')
                    elif model.pt and isinstance(model.model, SegmentationModel):
                        LOGGER.warning('WARNING ⚠️ YOLOv5 SegmentationModel is not yet AutoShape compatible. '
                                       'You will not be able to run inference with this model.')
                    else:
                        model = AutoShape(model)  # for file/URI/PIL/cv2/np inputs and NMS
            except Exception:
                model = attempt_load(path, device=device, fuse=False)  # arbitrary model
        else:
            cfg = list((Path(__file__).parent / 'models').rglob(f'{path.stem}.yaml'))[0]  # model.yaml path
            model = DetectionModel(cfg, channels, classes)  # create model
            if pretrained:
                ckpt = torch.load(attempt_download(path), map_location=device)  # load
                csd = ckpt['model'].float().state_dict()  # checkpoint state_dict as FP32
                csd = intersect_dicts(csd, model.state_dict(), exclude=['anchors'])  # intersect
                model.load_state_dict(csd, strict=False)  # load
                if len(ckpt['model'].names) == classes:
                    model.names = ckpt['model'].names  # set class names attribute
        if not verbose:
            LOGGER.setLevel(logging.INFO)  # reset to default
        return model.to(device)

    except Exception as e:
        help_url = 'https://github.com/ultralytics/yolov5/issues/36'
        s = f'{e}. Cache may be out of date, try `force_reload=True` or see {help_url} for help.'
        raise Exception(s) from e


def custom(path='path/to/model.pt', autoshape=True, _verbose=True, device=None):
    # YOLOv5 custom or local model
    return _create(path, autoshape=autoshape, verbose=_verbose, device=device)


def yolov5n(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-nano model https://github.com/ultralytics/yolov5
    return _create('yolov5n', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5s(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-small model https://github.com/ultralytics/yolov5
    return _create('yolov5s', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5m(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-medium model https://github.com/ultralytics/yolov5
    return _create('yolov5m', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5l(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-large model https://github.com/ultralytics/yolov5
    return _create('yolov5l', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5x(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-xlarge model https://github.com/ultralytics/yolov5
    return _create('yolov5x', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5n6(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-nano-P6 model https://github.com/ultralytics/yolov5
    return _create('yolov5n6', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5s6(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-small-P6 model https://github.com/ultralytics/yolov5
    return _create('yolov5s6', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5m6(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-medium-P6 model https://github.com/ultralytics/yolov5
    return _create('yolov5m6', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5l6(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-large-P6 model https://github.com/ultralytics/yolov5
    return _create('yolov5l6', pretrained, channels, classes, autoshape, _verbose, device)


def yolov5x6(pretrained=True, channels=3, classes=80, autoshape=True, _verbose=True, device=None):
    # YOLOv5-xlarge-P6 model https://github.com/ultralytics/yolov5
    return _create('yolov5x6', pretrained, channels, classes, autoshape, _verbose, device)


if __name__ == '__main__':
    import argparse
    from pathlib import Path

    import numpy as np
    from PIL import Image

    from yolov5.utils.general import cv2, print_args

    # Argparser
    parser = argparse.ArgumentParser()
    parser.add_argument('--model', type=str, default='yolov5s', help='model name')
    opt = parser.parse_args()
    print_args(vars(opt))

    # Model
    model = _create(name=opt.model, pretrained=True, channels=3, classes=80, autoshape=True, verbose=True)
    # model = custom(path='path/to/model.pt')  # custom

    # Images
    imgs = [
        'data/images/zidane.jpg',  # filename
        Path('data/_images/zidane.jpg'),  # Path
        'https://ultralytics.com/images/zidane.jpg',  # URI
        cv2.imread('data/_images/bus.jpg')[:, :, ::-1],  # OpenCV
        Image.open('data/_images/bus.jpg'),  # PIL
        np.zeros((320, 640, 3))]  # numpy

    # Inference
    results = model(imgs, size=320)  # batched inference

    # Results
    results.print()
    results.save()


Code/predict.py
import cv2
import torch
from torchvision import models
import torchvision.transforms as tvt
import torch.nn as nn
import matplotlib.pyplot as plt
import argparse
import os
import sys


def main():
    # Gets the user parameters
    parser = argparse.ArgumentParser(description='script to discriminate the presence or absence of the polar body')
    parser.add_argument('--input', help='path/to/input <image file>', type=str, required=True)
    parser.add_argument('--output', help='path/to/output <text file>', type=str, required=True)
    args = parser.parse_args()

    img_name = args.input
    txt_name = args.output
    base_name = os.path.basename(img_name).split('.')[0]

    out_dir = os.path.dirname(txt_name)
    if out_dir == '':
        out_dir = './'
        txt_name = os.path.join(out_dir, txt_name)

    # Sets the model weights
    det_weights_path = os.path.join(WEIGHT_DIR, DET_WEIGHT_FILE)
    clf_weights_path = os.path.join(WEIGHT_DIR, CLF_WEIGHT_FILE)

    # Checks the different src/dst paths
    if not os.path.exists(det_weights_path):
        sys.exit('Error: path to "%s" does not exist' % det_weights_path)

    if not os.path.exists(clf_weights_path):
        sys.exit('Error: path to "%s" does not exist' % clf_weights_path)

    if not os.path.exists(img_name):
        sys.exit('Error: path to "%s" does not exist' % img_name)

    if not os.path.exists(os.path.dirname(txt_name)):
        sys.exit('Error: path to "%s" does not exist' % txt_name)

    try:
        # Sets a device (CPU or GPU)
        device = torch.device('cuda:0') if torch.cuda.is_available() else torch.device('cpu')

        # Creates the detection model (YOLOv5)
        print('Detecting object...')
        det_model = torch.hub.load('.',
                                   'custom',
                                   path=det_weights_path,
                                   force_reload=True,
                                   source='local')

        # Puts the detection model in a device (CPU or GPU)
        det_model = det_model.to(device)

        # Initializes the detection model
        det_model.compute_iou = 0.6  # IoU threshold
        det_model.conf = 0.6  # Confidence threshold
        det_model.max_det = 1  # Max number of detections

        # Reads the input image
        img = cv2.imread(img_name)
        img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)

        '''
        # Converts the color image to grayscale format
        img = cv2.imread(img_name, cv2.IMREAD_GRAYSCALE)
        
        # Applies the CLAHE algorithm for contrast enhancement
        clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8, 8))
        img = clahe.apply(img)
        img = np.stack((img,) * 3, axis=-1)
        '''

        # Sets the detector to evaluation mode
        det_model.eval()

        # Performs object detection
        results = det_model(img)
        detections = results.xyxy[0].detach().cpu().numpy()

        # Checks if there is oocyte detections:
        if len(detections) > 0:
            box = detections[0][:4]
            x, y, w, h = round(box[0]), round(box[1]), round(box[2] - box[0]), round(box[3] - box[1])
        else:
            sys.exit('No oocyte detection')
        print('Done\n')

        # Creates the classification model
        clf_model = models.resnet101()

        # Adds a new final layer
        nr_filters = clf_model.fc.in_features  # Number of input features of last layer
        clf_model.fc = nn.Linear(nr_filters, N_OUTPUT_NEURONS)

        # Loads the trained model
        clf_model.load_state_dict(torch.load(clf_weights_path))

        # Puts the classification model in a device (CPU or GPU)
        clf_model = clf_model.to(device)

        # Sets the classifier to evaluation mode
        clf_model.eval()

        # Crops the image
        crop_img = img[y:y + h, x:x + w]

        # Composes the image transformations
        transform_img = tvt.Compose([tvt.ToPILImage(),
                                     tvt.Resize((IMG_HEIGHT, IMG_WIDTH), tvt.InterpolationMode.BILINEAR),
                                     tvt.Grayscale(num_output_channels=3),
                                     tvt.ToTensor()])

        # Applies the transformations to the cropped image
        crop_img = transform_img(crop_img)

        # Puts the transformed image in a device
        crop_img = torch.autograd.Variable(crop_img, requires_grad=False).to(device).unsqueeze(0)

        # Performs classification
        print('Discriminating the presence/absence of the polar body ...')
        with torch.no_grad():
            prd = clf_model(crop_img.float())

        # Gets the confidence and label tensors
        confidence = torch.sigmoid(prd)
        label = (confidence > 0.5).type(torch.uint8)

        # Puts the confidence and label tensors in numpy arrays and extracts the scalar values
        confidence = confidence.data.cpu().detach().numpy().squeeze(0)[0]
        label = label.data.cpu().detach().numpy().squeeze(0)[0]

        if label == 1:
            msg = 'the polar body is present'
        else:
            msg = 'the polar body is not present'

        # Saves the predicted label to a text file
        print('Confidence=%.2f\nLabel=%d\nResult: %s' % (confidence, label, msg))
        output = open(txt_name, "w")
        output.write(str(label))
        output.close()
        print('Done')

        # Saves the cropped image with a summary of the prediction
        fig, ax = plt.subplots()
        ax.imshow(img[y:y + h, x:x + w])
        ax.axis('off')
        ax.set_title('Confidence=%.2f Label=%d \nResult: %s' % (confidence, label, msg))
        fig.tight_layout()
        plt.savefig(os.path.join(out_dir, base_name + '_result.jpg'))
        plt.close('all')

    except Exception as e:
        print('Exception: %s' % str(e))


if __name__ == '__main__':
    # Main parameters
    IMG_WIDTH = 256
    IMG_HEIGHT = 256
    N_OUTPUT_NEURONS = 1
    WEIGHT_DIR = '../weights'
    DET_WEIGHT_FILE = 'oocyte_det.pt'
    CLF_WEIGHT_FILE = 'polar_body_clf.pt'

    # Call to main function
    main()


Code/predict_with_gradcam.py
import numpy as np
import cv2
import torch
import torch.nn as nn
from torchvision import models
import torchvision.transforms as tvt
import matplotlib.pyplot as plt
from matplotlib import colormaps
from torchvision.transforms.functional import to_pil_image
from PIL import Image
import argparse
import os
import sys

gradients = torch.Tensor()
activations = torch.Tensor()


def backward_hook(module, grad_input, grad_output):
    global gradients  # refers to the variable in the global scope
    print('Backward hook running...')
    gradients = grad_output
    # Prints the size of the gradients tensor
    # print('Gradients size: %s' % str(gradients[0].size()))


def forward_hook(module, args, output):
    global activations  # refers to the variable in the global scope
    print('Forward hook running...')
    activations = output
    # Prints the size of the activations tensor
    # print('Activations size: %s' % str(activations.size()))


def main():
    # Gets the user parameters
    parser = argparse.ArgumentParser(description='script to discriminate the presence or absence of the polar body')
    parser.add_argument('--input', help='path/to/input <image file>', type=str, required=True)
    parser.add_argument('--output', help='path/to/output <text file>', type=str, required=True)
    args = parser.parse_args()

    img_name = args.input
    txt_name = args.output
    base_name = os.path.basename(img_name).split('.')[0]

    out_dir = os.path.dirname(txt_name)
    if out_dir == '':
        out_dir = './'
        txt_name = os.path.join(out_dir, txt_name)

    # Sets the model weights
    det_weights_path = os.path.join(WEIGHT_DIR, DET_WEIGHT_FILE)
    clf_weights_path = os.path.join(WEIGHT_DIR, CLF_WEIGHT_FILE)

    # Checks the different src/dst paths
    if not os.path.exists(det_weights_path):
        sys.exit('Error: path to "%s" does not exist' % det_weights_path)

    if not os.path.exists(clf_weights_path):
        sys.exit('Error: path to "%s" does not exist' % clf_weights_path)

    if not os.path.exists(img_name):
        sys.exit('Error: path to "%s" does not exist' % img_name)

    if not os.path.exists(os.path.dirname(txt_name)):
        sys.exit('Error: path to "%s" does not exist' % txt_name)

    try:
        # Sets a device (CPU or GPU)
        device = torch.device('cuda:0') if torch.cuda.is_available() else torch.device('cpu')

        # Creates the detection model (YOLOv5)
        print('Detecting object...')
        det_model = torch.hub.load('.',
                                   'custom',
                                   path=det_weights_path,
                                   force_reload=True,
                                   source='local')

        # Puts the detection model in a device (CPU or GPU)
        det_model = det_model.to(device)

        # Initializes the detection model
        det_model.compute_iou = 0.6  # IoU threshold
        det_model.conf = 0.6  # Confidence threshold
        det_model.max_det = 1  # Max number of detections

        # Reads the input image
        img = cv2.imread(img_name)
        img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
        # img = cv2.imread(img_name, cv2.IMREAD_GRAYSCALE)
        # img = np.stack((img,) * 3, axis=-1)

        # Sets the detector to evaluation mode
        det_model.eval()

        # Performs object detection
        results = det_model(img)
        detections = results.xyxy[0].detach().cpu().numpy()

        # Checks if there is oocyte detections:
        if len(detections) > 0:
            box = detections[0][:4]
            x, y, w, h = round(box[0]), round(box[1]), round(box[2] - box[0]), round(box[3] - box[1])
        else:
            sys.exit('No oocyte detection')
        print('Done\n')

        # Creates the classification model (ResNet101)
        clf_model = models.resnet101()

        # Adds a new final layer
        nr_filters = clf_model.fc.in_features  # Number of input features of last layer
        clf_model.fc = nn.Linear(nr_filters, N_OUTPUT_NEURONS)

        # Loads the trained model for classification
        clf_model.load_state_dict(torch.load(clf_weights_path))

        # Puts the classification model in a device (CPU or GPU)
        clf_model = clf_model.to(device)

        # Sets the classifier to evaluation mode
        clf_model.eval()

        # Crops the image
        crop_img = img[y:y + h, x:x + w]
        orig_crop = crop_img.copy()

        # Composes the image transformations
        transform_img = tvt.Compose([tvt.ToPILImage(),
                                     tvt.Resize((IMG_HEIGHT, IMG_WIDTH), tvt.InterpolationMode.BILINEAR),
                                     tvt.Grayscale(num_output_channels=3),
                                     tvt.ToTensor()])

        # Applies the transformations to the cropped image
        crop_img = transform_img(crop_img)

        # Puts the transformed image in a device
        crop_img = torch.autograd.Variable(crop_img, requires_grad=False).to(device).unsqueeze(0)

        # Executes the classifier to discriminate the presence or absence of polar body
        print('Discriminating the presence/absence of the polar body ...')

        # ------- CODE BLOCK BEGIN ------- #
        # The GradCam algorithm is used to "explain" the model predictions

        # Gets the last convolutional block
        target_layers = clf_model.layer4[-1]

        # Registers a backward hook on the module
        b_hook = target_layers.register_full_backward_hook(backward_hook, prepend=False)

        # Registers a forward hook on the module
        f_hook = target_layers.register_forward_hook(forward_hook, prepend=False)

        # Performs the forward/backward steps in a single call
        clf_model(crop_img.float()).backward()

        # Pools the gradients across the channels
        print(gradients[0].shape)
        pooled_gradients = torch.mean(gradients[0], dim=[0, 2, 3])

        # Weights the channels by corresponding gradients
        for i in range(activations.size()[1]):
            activations[:, i, :, :] *= pooled_gradients[i]

        # Averages the channels of the activations
        heatmap = torch.mean(activations, dim=1)

        # Applies Relu on top of the 'heatmap'
        # heatmap = nn.functional.relu(heatmap)

        # Normalizes the heatmap
        heatmap /= torch.max(heatmap)

        # Gets the model prediction
        prd = clf_model(crop_img.float())
        confidence = torch.sigmoid(prd).cpu().detach().numpy()
        label = (confidence > 0.5).astype(np.uint8)

        if label == 1:
            msg = 'the polar body is present'
        else:
            msg = 'the polar body is not present'

        # Creates a figure and plots the crop
        fig, ax = plt.subplots()
        ax.axis('off')  # removes the axis markers
        ax.imshow(orig_crop)

        # Resize the heatmap to the same size as the crop
        overlay = to_pil_image(
            heatmap.cpu().detach(),
            mode='F').resize((orig_crop.shape[0], orig_crop.shape[1]),
                             resample=Image.Resampling.BICUBIC)

        # Applies the 'jet' colormap
        cmap = colormaps['jet']
        overlay = (255 * cmap(np.asarray(overlay) ** 2)[:, :, :3]).astype(np.uint8)

        # Plots the heatmap on the same axes but with alpha < 1 (this defines the transparency of the heatmap)
        ax.axis('off')
        ax.set_title('Confidence=%.4f Label=%d \nResult: %s' % (confidence, label, msg))
        ax.imshow(overlay, alpha=0.4, interpolation='nearest', extent=(0, orig_crop.shape[1], orig_crop.shape[0], 0))

        # Saves the plot
        plt.tight_layout()
        plt.savefig(os.path.join(out_dir, base_name + '_result.jpg'))

        # Removes the hooks
        f_hook.remove()
        b_hook.remove()
        # ------- CODE BLOCK END ------- #

        print('Confidence=%.4f\nLabel=%d\nResult: %s' % (confidence, label, msg))
        output = open(txt_name, "w")
        output.write(str(label))
        output.close()
        print('Done')

    except Exception as e:
        print('Exception: %s' % str(e))


if __name__ == '__main__':
    # Main parameters
    IMG_WIDTH = 256
    IMG_HEIGHT = 256
    N_OUTPUT_NEURONS = 1
    WEIGHT_DIR = '../weights'
    DET_WEIGHT_FILE = 'oocyte_det.pt'
    CLF_WEIGHT_FILE = 'polar_body_clf.pt'

    # Call to main function
    main()


Code/train_classifier.py
import numpy as np
import torch
from torch.utils.data import DataLoader
from torchvision import transforms as tvt
from torchvision import datasets
from torchvision import models
import pandas as pd
import matplotlib.pyplot as plt
import os
import sys
from tqdm import tqdm
from time import sleep


def fit(model, device, train_dataloader, val_dataloader, epochs=100, lr=1e-3):
    """
    Function to train the model and save the best and the last weights under a selected metric
    Args:
        model (torch model): NN model
        device (torch device): device (CPU or GPU)
        train_dataloader (torch dataloader): loads images for training
        val_dataloader (torch dataloader): loads images for validation
        epochs (int): number of epochs for training
        lr (float): initial learning rate

    Returns:
        a dictionary with the history training
    """

    # Loss functions
    criterion = torch.nn.BCEWithLogitsLoss()

    # Optimizer
    optimizer = torch.optim.AdamW(params=model.parameters(), lr=lr, weight_decay=1e-6, amsgrad=False)
    # optimizer = torch.optim.Adam(params=model.parameters(), lr=lr, weight_decay=1e-5, amsgrad=False)

    # Learning rate scheduler
    scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
        optimizer,
        mode='min',
        factor=0.5,
        patience=5,
        verbose=True,
        min_lr=1e-6
    )

    # scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=30, gamma=0.1)

    # Puts the model in a device (CPU or GPU)
    model.to(device)

    # Training history dictionary
    hist = {'loss': [], 'accuracy': [], 'val_loss': [], 'val_accuracy': []}

    # Initial validation loss reference
    # val_loss_ref = 1.

    # Initial validation accuracy reference
    val_acc_ref = 0

    # Training loop
    for epoch in range(1, epochs + 1):
        train_total_correct = 0
        train_total_samples = 0

        val_total_correct = 0
        val_total_samples = 0

        # Loads images and labels
        bar = tqdm(train_dataloader)

        # Initializes the loss and IoU values for training
        train_loss, train_acc = [], []

        # Sets the model to train mode
        model.train()

        # Training loop
        for images, labels in bar:
            # Puts the images and the labels in the device
            images, labels = images.to(device), labels.to(device)

            # Add a dim to the labels tensor
            labels = torch.unsqueeze(labels, -1)

            # Sets the gradients to None
            optimizer.zero_grad()

            # Model predictions
            predictions = model(images)

            # Computes the training loss
            loss = criterion(predictions, labels.float())

            # Performs the backpropagation
            loss.backward()

            # Performs the parameter update
            optimizer.step()

            # Converts predictions to binary values
            # predictions = (predictions > 0.5).type(torch.uint8)
            predictions = (torch.sigmoid(predictions) > 0.5).type(torch.uint8)

            # Updates the running total of correct predictions and samples
            train_total_correct += (predictions == labels).sum().item()
            train_total_samples += labels.size(0)

            # Calculates the accuracy for this epoch
            accuracy = train_total_correct / train_total_samples

            # Stores the training parameters for the current batch
            train_loss.append(loss.item())
            train_acc.append(accuracy)
            bar.set_description("loss=%.5f acc=%.5f" % (float(np.mean(train_loss)), float(np.mean(train_acc))))
            sleep(0.1)

        # Stores the training parameters for the current epoch
        hist['loss'].append(np.mean(train_loss))
        hist['accuracy'].append(np.mean(train_acc))

        # Loads images and labels for validation
        bar = tqdm(val_dataloader)

        # Initializes the loss and IoU values for validation
        val_loss, val_acc = [], []

        # Sets the model to evaluation mode
        model.eval()

        # Validation loop
        with torch.no_grad():
            for images, labels in bar:
                # Puts the images and the labels in the device
                images, labels = images.to(device), labels.to(device)

                # Add a final dim to the labels tensor
                labels = torch.unsqueeze(labels, -1)

                # Model predictions
                predictions = model(images)

                # Computes the validation loss
                loss = criterion(predictions, labels.float())

                # Converts predictions to binary
                # predictions = (predictions > 0.5).type(torch.uint8)
                predictions = (torch.sigmoid(predictions) > 0.5).type(torch.uint8)

                # Updates the running total of correct predictions and samples
                val_total_correct += (predictions == labels).sum().item()
                val_total_samples += labels.size(0)

                # Calculates the validation accuracy for this epoch
                accuracy = val_total_correct / val_total_samples

                # Stores the validation parameters for the current batch
                val_loss.append(loss.item())
                val_acc.append(accuracy)
                bar.set_description("val_loss=%.5f val_acc=%.5f" % (float(np.mean(val_loss)), float(np.mean(val_acc))))
                sleep(0.1)

        # Stores the validation parameters for the current epoch
        hist['val_loss'].append(np.mean(val_loss))
        hist['val_accuracy'].append(np.mean(val_acc))

        # Prints the results at ending the epoch
        print('Epoch %d/%d loss=%.5f acc=%.5f val_loss=%.5f val_acc=%.5f' %
              (epoch,
               epochs,
               float(np.mean(train_loss)),
               float(np.mean(train_acc)),
               float(np.mean(val_loss)),
               float(np.mean(val_acc))))

        # Saves the best model (using validation accuracy criteria)
        # if np.mean(val_loss) < val_loss_ref:
        if np.mean(val_acc) > val_acc_ref:
            print('Saving model...')
            torch.save(model.state_dict(), os.path.join(RESULTS_DIR, BEST_WEIGHT_FILE))
            # val_loss_ref = np.mean(val_loss)
            val_acc_ref = np.mean(val_acc)

        # Applies the learning rate scheduler policy
        scheduler.step(np.mean(val_loss))
        print()
        sleep(0.1)

    # Saves the last model
    print('Saving model...')
    torch.save(model.state_dict(), os.path.join(RESULTS_DIR, LAST_WEIGHT_FILE))
    return hist


def main():
    """
    Classes:
        0- Not Mature
        1- Mature
    """

    # Creates a directory for the current experiment (if the directory exists, the experiment is aborted)
    try:
        if not os.path.exists(RESULTS_DIR):
            os.makedirs(RESULTS_DIR)
        else:
            sys.exit('Path to %s already exists. Please, assign another directory for the experiment results' %
                     RESULTS_DIR)
    except Exception as e:
        sys.exit('Exception: %s' % str(e))

    # Composes the image transformations for training/validation
    train_transform_img = tvt.Compose([
        # tvt.ToPILImage(),
        tvt.Resize((HEIGHT, WIDTH), tvt.InterpolationMode.BILINEAR),
        tvt.ColorJitter(brightness=0.2, contrast=0.2, saturation=0., hue=0.),
        tvt.Grayscale(num_output_channels=3),
        tvt.RandomHorizontalFlip(p=0.5),
        tvt.RandomVerticalFlip(p=0.5),
        tvt.ToTensor()])

    val_transform_img = tvt.Compose([
        # tvt.ToPILImage(),
        tvt.Resize((HEIGHT, WIDTH), tvt.InterpolationMode.BILINEAR),
        tvt.Grayscale(num_output_channels=3),
        tvt.ToTensor()])

    try:

        # Gets the train/validation datasets
        train_data = datasets.ImageFolder(os.path.join(DATA_DIR, TRAIN_DIR), transform=train_transform_img)
        val_data = datasets.ImageFolder(os.path.join(DATA_DIR, VAL_DIR), transform=val_transform_img)

        # Gets the dataloaders
        train_loader = torch.utils.data.DataLoader(train_data, shuffle=True, batch_size=BATCH_SIZE, pin_memory=True)
        val_loader = torch.utils.data.DataLoader(val_data, shuffle=False, batch_size=BATCH_SIZE, pin_memory=True)

        # Sets the device (CPU or GPU)
        device = 'cuda:0' if torch.cuda.is_available() else 'cpu'

        # Loads a ResNet101 model with default weights (IMAGENET1K_V2)
        model = models.resnet101(weights=models.ResNet101_Weights.DEFAULT)
        num_features = model.fc.in_features
        model.fc = torch.nn.Linear(num_features, N_OUTPUT_NEURONS)

        # Trains the model
        hist = fit(model, device, train_loader, val_loader, epochs=N_EPOCHS, lr=LR)
        df = pd.DataFrame(hist)
        df.to_csv(os.path.join(RESULTS_DIR, CSV_FILE), sep=',', index=False, encoding='latin1')
        title = 'Model: %s  Loss function: %s\n Input image format: %s' % \
                (MODEL_NAME, LOSS_FUNCTION_NAME, IMG_FORMAT)
        df.plot(grid=True, title=title, xlabel='Epoch', ylabel='Loss / Accuracy', figsize=(10, 8))
        plt.tight_layout()
        plt.savefig(os.path.join(RESULTS_DIR, IMG_FILE))
    except Exception as e:
        sys.exit('Exception: %s' % str(e))


if __name__ == '__main__':
    torch.set_float32_matmul_precision('medium')

    # Main parameters
    MODEL_NAME = 'ResNet101'
    LOSS_FUNCTION_NAME = 'SoftBCEWithLogits'
    IMG_FORMAT = 'RGB'
    DATA_DIR = '../data'
    WEIGHT_DIR = '../weights'
    BEST_WEIGHT_FILE = 'best.pt'
    LAST_WEIGHT_FILE = 'last.pt'
    RUN_DIR = 'runs'
    TRAIN_DIR = 'train'
    VAL_DIR = 'val'
    CSV_FILE = 'train.csv'
    IMG_FILE = 'train.jpg'
    WIDTH = 256  # image width
    HEIGHT = 256  # image height
    LR = 1e-4
    BATCH_SIZE = 10
    N_EPOCHS = 50
    N_OUTPUT_NEURONS = 1
    N_EXP = 1
    RESULTS_DIR = os.path.join(RUN_DIR, TRAIN_DIR, str(N_EXP))
    # SEED = 42
    # np.random.seed(SEED)
    # torch.manual_seed(SEED)
    # torch.cuda.manual_seed(SEED)

    # Main function call
    main()
